wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1064 @@
+        uint8_t fs_msb, fs_lsb;
+        hex_to_bytes(filesz_s + 4, &fs_msb, 1);
+        hex_to_bytes(filesz_s + 6, &fs_lsb, 1);
+        file_sz = (fs_msb << 8) + fs_lsb;
+        WOLFSSL_MSG("Stat successful on file");
+    }
+
+    if (file_sz > max_size)
+    {
+        WOLFSSL_MSG("iotsafe_readfile: insufficient space in buffer");
+        return -1;
+    }
+
+    while (off < file_sz) {
+        byte off_p1, off_p2;
+        off_p1 = (off >> 8) & 0xff;
+        off_p2 = (off & 0xff);
+        iotsafe_cmd_start(csim_cmd, IOTSAFE_CLASS, IOTSAFE_INS_READ_FILE,
+                off_p1, off_p2);
+        iotsafe_cmd_add_tlv(csim_cmd, IOTSAFE_TAG_FILE_ID, 1, &file_id);
+        iotsafe_cmd_complete(csim_cmd);
+        ret = expect_csim_response(csim_cmd, (word32)strlen(csim_cmd), &resp);
+        if (ret > 0) {
+            if (ret > 2 * 0xF3)
+                ret = 2 * 0xF3;
+            if (ret > 2 * (file_sz - off))
+                ret = 2 * (file_sz - off);
+            hex_to_bytes(resp, content + off, (ret / 2));
+            off += ret/2;
+        } else {
+            WOLFSSL_MSG("IoTSafe: Error reading file.");
+            return -1;
+        }
+    }
+    return off;
+}
+
+static int iotsafe_getrandom(unsigned char* output, unsigned long sz)
+{
+    char *resp = NULL;
+    int ret;
+    byte len = (byte)sz;
+    if ( !wolfIoT_initialized)
+    {
+        iotsafe_init();
+    }
+    iotsafe_cmd_start(csim_cmd, IOTSAFE_CLASS, IOTSAFE_INS_GETRANDOM,0,0);
+    bytes_to_hex(&len, csim_cmd + AT_CSIM_CMD_SIZE + 8, 1);
+    csim_cmd[AT_CSIM_CMD_SIZE + AT_CMD_HDR_SIZE] = 0;
+    iotsafe_cmd_complete(csim_cmd);
+    if (expect_csim_response(csim_cmd, (word32)GETRAND_CMD_SIZE, &resp) < 1) {
+        WOLFSSL_MSG("Unexpected reply from RAND");
+        return WC_HW_E;
+    }
+    ret = hexbuffer_conv(resp, output, sz);
+    if (ret < 0)
+        return ret;
+    expect_tok(NULL, 0, NULL, NULL);
+    return 0; /* WolfSSL API expects 0 */
+}
+
+
+
+#ifdef HAVE_ECC
+
+#define IOTSAFE_ECC_KSIZE 32
+
+static int iotsafe_gen_keypair(byte wr_slot)
+{
+    char *resp;
+    iotsafe_cmd_start(csim_cmd, IOTSAFE_CLASS, IOTSAFE_INS_GEN_KEYPAIR, 0, 0);
+    iotsafe_cmd_add_tlv(csim_cmd, IOTSAFE_TAG_PRIVKEY_ID, 1, &wr_slot);
+    iotsafe_cmd_complete(csim_cmd);
+    if (expect_csim_response(csim_cmd, (word32)strlen(csim_cmd), &resp) < 1) {
+        WOLFSSL_MSG("Unexpected reply from Keygen");
+        return WC_HW_E;
+    }
+    return 0;
+}
+
+static int iotsafe_get_public_key(byte idx, ecc_key *key)
+{
+    int ret;
+    char *resp;
+    char *rkey, *ktype, *payload_str;
+    char Qx[IOTSAFE_ECC_KSIZE * 2 + 1], Qy[IOTSAFE_ECC_KSIZE * 2 + 1];
+
+    /* exporting generated public key */
+    iotsafe_cmd_start(csim_cmd, IOTSAFE_CLASS, IOTSAFE_INS_READ_KEY,0, 0);
+    iotsafe_cmd_add_tlv(csim_cmd, IOTSAFE_TAG_PUBKEY_ID, 1, &idx);
+    iotsafe_cmd_complete(csim_cmd);
+    ret = expect_csim_response(csim_cmd, (word32)strlen(csim_cmd), &resp);
+    if (ret < 1) {
```

Comment:
I would prefer a single return point. Could the -1 return error codes be improved?

Reply:
Refactored all these functions to have single return point. Improved most  '-1' errors.